AZ3166/src/libraries/AzureIoT/src/OTAUtils.cpp

```diff
@@ -0,0 +1,144 @@
+    char *serialized_string = NULL;
+    json_object_set_string(firmware_object, "currentFwVersion", currentFwVersion);
+    json_object_set_string(firmware_object, "fwUpdateStatus", fwUpdateStatus);
+    if (pendingFwVersion != NULL) {
+        json_object_set_string(firmware_object, "pendingFwVersion", pendingFwVersion);
+    }
+    if (fwUpdateSubstatus != NULL) {
+        json_object_set_string(firmware_object, "fwUpdateSubstatus", fwUpdateSubstatus);
+    }
+    if (lastFwUpdateStartTime != NULL) {
+        json_object_set_string(firmware_object, "lastFwUpdateStartTime", lastFwUpdateStartTime);
+        json_object_set_null(firmware_object, "lastFwUpdateEndTime");
+    }
+    if (lastFwUpdateEndTime != NULL) {
+        json_object_set_string(firmware_object, "lastFwUpdateEndTime", lastFwUpdateEndTime);
+    }
+    JSON_Value *root_value = json_value_init_object();
+    JSON_Object *root_object = json_value_get_object(root_value);
+    json_object_set_value(root_object, "firmware", firmware_value);
+    serialized_string = json_serialize_to_string_pretty(root_value);
+    json_value_free(root_value);
+    return DevKitMQTTClient_ReportState(serialized_string);
+}
+
+void ota_callback(const unsigned char *payLoad, size_t size) {
+    fw_info_free_string(otaFwInfo);
+    char *deviceTwinPayLoad = (char *)malloc(size + 1);
+    if (deviceTwinPayLoad != NULL)
+    {
+        memcpy(deviceTwinPayLoad, payLoad, size);
+        deviceTwinPayLoad[size] = '\0';
+    } else {
+        free(deviceTwinPayLoad);
+        return;
+    }
+    JSON_Value *root_value;
+    root_value = json_parse_string(deviceTwinPayLoad);
+    if (json_value_get_type(root_value) != JSONObject)
+    {
+        if (root_value != NULL)
+        {
+            json_value_free(root_value);
+        }
+        free(deviceTwinPayLoad);
+        LogError("parse %s failed", deviceTwinPayLoad);
+        return;
+    }
+    free(deviceTwinPayLoad);
+    JSON_Object *root_object = json_value_get_object(root_value);
+    JSON_Object *desired_object = json_object_get_object(root_object, "desired");
+    desired_object = desired_object == NULL ? root_object : desired_object;
+    if (desired_object != NULL)
+    {
+        JSON_Object* firmwareVal = json_object_get_object(desired_object, "firmware");
+        if (firmwareVal != NULL) {
+            LogInfo("Get firmware value from device twin.");
+            otaFwInfo.fwVersion = strdup(json_object_get_string(firmwareVal, "fwVersion"));
+            otaFwInfo.fwPackageURI = strdup(json_object_get_string(firmwareVal, "fwPackageURI"));
+            otaFwInfo.fwPackageCheckValue = strdup(json_object_get_string(firmwareVal, "fwPackageCheckValue"));
+            otaFwInfo.fwSize = json_object_get_number(firmwareVal, "fwSize");
+            if (otaFwInfo.fwVersion != NULL && otaFwInfo.fwPackageURI != NULL && otaFwInfo.fwPackageCheckValue != NULL) {
+                json_value_free(root_value);
```

Comment:
no need to free root and return here. But you can fw_info_free_string(otaFwInfo) here if you want to make sure all field align.